src/utils/actionToUrl.js

```diff
@@ -3,40 +3,40 @@ import { compileUrl, cleanBasename } from './index'
 
 import type {
   Route,
-  Payload,
-  Params,
-  RoutesMap,
+  Routes,
   ReceivedAction as Action,
   Options
 } from '../flow-types'
 
 export default (
   action: Action,
   api: Object,
-  prevRoute?: Object
-): string => {
-  const { routes, options: opts } = api
-  const { type, params, query, state, hash, basename } = action
+  prevRoute?: string
+): Object => {
+  const { routes, options: opts }: Object = api
+  const { type, params, query, state, hash, basename }: Action = action
 
   const route = routes[type] || {}
-  const path = typeof route === 'object' ? route.path : route
+  const path: string | void | string | void = typeof route === 'object' ? route.path : route
 
-  const p = formatParams(params, route, opts)
+  const p: void | {} = formatParams(params, route, opts)
   const q = formatQuery(query, route, opts)
-  const s = formatState(state, route, opts)
-  const h = formatHash(hash, route, opts)
+  const s: ?Object = formatState(state, route, opts)
```

Comment:
formatState always returns an `Object` (see how it has a default value of an object for its `state` arg). Im pretty sure the idea is none of these return undefined. Overall, that's the goal as much as possible--so we dont have to write lots of conditional logic to satisfy the flow compiler if undefined/null valuables are possible. 

So to re-iterate, the goal was to write as much of this as possible to avoid undefined/null values with default objects (or strings). There may be exceptions--so look out for my usage of default values, as that pretty much guarantees undefined/null values are replaced with something as early on in the execution as possible.

Reply:
Okay yeah that makes sense. 

So to confirm, you want me to replace/look out for things like this and try to always state a defined value. 

In this scenario “Object” and not “?Object”

Reply:
exactly. where possible.

Reply:
Oh perfect! Dude I love that shit. Nothing is worse than tracking down a damn missing type 😂😂

I’m a huge fan of PropTypes because of that. So pretty much I’ll do the same thing. Just needed clarification